### PR TITLE
New eigenvector gradient regularization scheme

### PR DESCRIPTION
**Pull request by mfschubert**
Motivated by https://github.com/facebookresearch/fmmax/issues/105, a new regularization scheme.

- Updates the regularization scheme to use an eps value which depends upon the range of eigenvalues.
- Adds some documentation which references approaches that various papers have used.
- Adds tests which compare gradients to those computed by `jnp.linalg.eigh`, which is differentiable "out of the box". The tests include matrices with shifted and scaled eigenvalues, which would fail with the previous regularization scheme.

@elinorbgr could you test this?
@smartalecH good to get your eyes on this also

### CODE REVIEW

File: src/fmmax/utils.py [hunk 1]
```diff
@@ -142,27 +145,57 @@ def _eig_cpu(matrix: jnp.ndarray) -> Tuple[jnp.ndarray, jnp.ndarray]:
...
+    # be identical, except scaled by 1 / 1e-6.)
+    #
+    # Therefore, we use Lorentzian broadening similar to torcwa, but with an eps
+    # value that is computed in a way that considers the eigenvalue range.
     eigenvalues_i = eigenvalues[..., jnp.newaxis, :]
     eigenvalues_j = eigenvalues[..., :, jnp.newaxis]
     delta_eig = eigenvalues_i - eigenvalues_j
+    eig_range = jnp.amax(jnp.abs(delta_eig), axis=(-2, -1), keepdims=True)
+    eps = jnp.maximum(eps_relative * eig_range, EIG_EPS_MINIMUM)
     f_broadened = delta_eig.conj() / (jnp.abs(delta_eig) ** 2 + eps)
```
Comment (elinorbgr) [inline src/fmmax/utils.py:199]: If the aim is to make the regularization scheme invariant to a global scale factor of the matrix, I think `eps` should be made proportional to the *square* of the absolute value of `delta_eig` instead, no?

Comment (mfschubert) [reply src/fmmax/utils.py]: Ah yes, good catch. I just adjusted the calculation to have eps depend on the square of the maximum eigenvalue difference. Tests still pass so I haven't touched the `EIG_EPS` module constant.

**Comment by elinorbgr** [comment]
So, I have not done extensive testing (and doing so would likely require me to retune all my optimizer hyperparameters), but from the light testing I've done, this new scheme (including the latest commit) does seem to behave quite well on my setup, possibly even better than what I was getting before by setting $\epsilon = 10^{-20}$ :+1: .

**Review by smartalecH — APPROVED** [review]
Sorry for the delay -- been on travel.

This looks great! Thank you both for the rigorous testing offline.
